File: 3653-xor-after-range-multiplication-queries-i/3653-xor-after-range-multiplication-queries-i.cpp
class Solution {
public:
    int xorAfterQueries(vector<int>& n, vector<vector<int>>& q) {
        int mod = 1e9 + 7;

        for (auto &a : q) {
            for (int i = a[0]; i <= a[1]; i += a[2]) {
                long long p = 1LL * n[i] * a[3];
                n[i] = (int)(p % mod);
            }
        }

        int s = 0;
        for (int x : n) s ^= x;

        return s;
    }
};

File: 3653-xor-after-range-multiplication-queries-i/3653-xor-after-range-multiplication-queries-i.py
import numpy as np
class Solution:
    def xorAfterQueries(self, nums: List[int], queries: List[List[int]]) -> int:
        arr = np.array(nums, dtype=np.int64)
        MOD = 10**9 + 7
        for li, ri, ki, vi in queries:
            arr[li : ri + 1 : ki] = (arr[li : ri + 1 : ki] * vi) % MOD
        return int(np.bitwise_xor.reduce(arr))